src/Inputs/Slider.tsx

```diff
@@ -1,79 +1,384 @@
+    rightValue: number;
+}
 
 export interface SliderProps extends LabelLayoutProps {
+    values?: Values;
+    onChange?: Function;
+
+    disabled?: boolean;
+    hasTwoKnobs?: boolean;
+    hasRail?: boolean;
+    hasPopup?: boolean;
+    step?: number;
+
     max?: number;
     min?: number;
-    step?: number;
-    disabled?: boolean;
-    value?: number;
-    trackColor?: string;
-    buttonColor?: string;
+
+    marks?: MarkProps[];
+    popupLeft?: number;
+    popupTop?: number;
+    popupWidth?: number;
+    popupHeight?: number;
+    theme?: DefaultTheme;
 }
 
 export const Slider: React.FunctionComponent<SliderProps> = ({
+    onChange = (): void => {},
+    hasPopup = false,
+    disabled = false,
+    hasRail = false,
+    hasTwoKnobs = false,
+    step = 1,
+    min = 0,
+    max = 100,
+    marks,
+    values = { leftValue: min, rightValue: max },
+    popupLeft = -12,
+    popupTop = -51,
+    popupWidth,
+    popupHeight = 20,
+    theme,
     ...props
-}): React.ReactElement => (
-    <LabelLayout {...props}>
-        <SliderElement {...props} type="range" />
-    </LabelLayout>
-);
-
-const SliderElement = styled.input<SliderProps>`
-    -webkit-appearance: none;
-    margin: 18px 0;
-    width: 100%;
-    // styles for slider adapted from: https://css-tricks.com/styling-cross-browser-compatible-range-inputs-css/
-    &:disabled {
-        cursor: not-allowed;
-        opacity: 0.6;
-    }
-    &:focus {
-        outline: none;
-    }
-    &::-webkit-slider-runnable-track {
-        width: 100%;
-        height: 6px;
-        cursor: pointer;
-        animate: 0.2s;
-        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
-        background: #3071a9;
-        border-radius: 10px;
-        border: 0.2px solid ${(props): string => props.trackColor || '#010101'};
-    }
-    &::-webkit-slider-thumb {
-        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
-        border: 1px solid #000000;
-        height: 16px;
-        width: 16px;
-        border-radius: 50%;
-        background: ${(props): string => props.buttonColor || '#ffffff'};
-        cursor: pointer;
-        -webkit-appearance: none;
-        margin-top: -5px;
-    }
-    &:focus::-webkit-slider-runnable-track {
-        background: #367ebd;
-    }
-    &::-moz-range-track {
+}): React.ReactElement => {
+    // DOM Elements
+    const barRef = useRef<HTMLDivElement>(null);
+    const selectedBarRef = useRef<HTMLDivElement>(null);
+    const marksBarRef = useRef<HTMLDivElement>(null);
+
+    // Knob Positions
+    const [rightKnobPosition, setRightKnobPosition] = useState(
+        values.rightValue,
+    );
+
+    const [leftKnobPosition, setLeftKnobPosition] = useState(values.leftValue);
+
+    const [isRightKnobDragging, setIsRightKnobDragging] = useState(false);
+    const [isLeftKnobDragging, setIsLeftKnobDragging] = useState(false);
+
+    const maxAndMinDifference = useMemo((): number => {
+        return max - min;
+    }, [max, min]);
```

Comment:
Does this need to be in a use Memo? why not just `const difference = max - min`

Reply:
@ryanarine You don't want to re-calculate the difference everytime one of the other props changes though.

Reply:
@ralph-dev The overhead of useMemo probably outweighs the time it takes to subtract two numbers

Reply:
I agree but in that case just have the function without useMemo at all? You never want things hanging in the component because the calculation will execute almost ever re-render essentially because it's a funtional component.

Reply:
Regardless if its a function or not, wouldn't the calculation execute every re-render? (with useMemo the calculation wouldn't execute every re-render but the useMemo itself would)

Does it matter that this is a functional component? if it was class component , the calculation would have to be in the render function(I think) which would execute every re-render (unless you have shouldComponentUpdate)

Reply:
@ryanarine Not really because if it was just a function then on a re-render it would just be a function creation not a calculation. UseMemo only helps when someone calls this function to not do the calculation again unless 2 variables changes